### PR TITLE
[SYCL] Fix possible failure when enqueing only a single host-task

### PR DESCRIPTION
**Pull request by s-kanaev**
The issue results in either segfault due to some nullptr dereference or in a deadlock when the buffer gets destroyed and the last access to it was through host task.

### CODE REVIEW

File: sycl/test/host-interop-task/host-task-failure.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    auto DstA = Dst.template get_access<mode::write>(CGH);
+
+    CGH.codeplay_host_task([=]() {
+      for (size_t Idx = 0; Idx < SrcA.get_count(); ++Idx)
+        DstA[Idx] = SrcA[Idx];
+    });
+  });
+}
+
+template <typename DataT, DataT B1Init, DataT B2Init>
```
Comment (v-klochkov) [inline sycl/test/host-interop-task/host-task-failure.cpp:34]: B1Init and B2init are unused.

File: sycl/test/host-interop-task/host-task-failure.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  });
+}
+
+void test() {
+  static constexpr int COUNT = 4;
+  queue Q;
+  buffer<int, 1> Buffer1{BUFFER_SIZE};
+  buffer<int, 1> Buffer2{BUFFER_SIZE};
+
+  // init the buffer with a'priori invalid data
```
Comment (v-klochkov) [inline sycl/test/host-interop-task/host-task-failure.cpp:53]: why -1 and -2 are ' a'priori invalid data' ? Those are just regular/good int values.

File: sycl/test/host-interop-task/host-task-failure.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    CGH.parallel_for<Init<DataT>>(BUFFER_SIZE, [=](item<1> Id) {
+      Acc1[Id] = -1;
+      Acc2[Id] = -2;
+    });
+  });
+}
+
+void test() {
+  static constexpr int COUNT = 4;
```
Comment (v-klochkov) [inline sycl/test/host-interop-task/host-task-failure.cpp:48]: COUNT is unused.

File: sycl/test/host-interop-task/host-task-failure.cpp (hunk 1 above)
Comment (s-kanaev) [reply sycl/test/host-interop-task/host-task-failure.cpp]: Removed.

File: sycl/test/host-interop-task/host-task-failure.cpp (hunk 2 above)
Comment (s-kanaev) [reply sycl/test/host-interop-task/host-task-failure.cpp]: Sorry, that's a rudiment comment from another test.

Comment (s-kanaev) [reply sycl/test/host-interop-task/host-task-failure.cpp]: Removed.

File: sycl/test/host-interop-task/host-task-failure.cpp (hunk 3 above)
Comment (s-kanaev) [reply sycl/test/host-interop-task/host-task-failure.cpp]: Removed.

File: sycl/test/host-interop-task/host-task-failure.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/backend/opencl.hpp>
+#include <CL/sycl/detail/cl.h>
```
Comment (sergey-semenov) [inline sycl/test/host-interop-task/host-task-failure.cpp:8]: Why are these includes needed here?

Comment (s-kanaev) [reply sycl/test/host-interop-task/host-task-failure.cpp]: They're unneeded here.
Removed.

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 5]
```diff
@@ -156,18 +156,33 @@ void Scheduler::cleanupFinishedCommands(EventImplPtr FinishedEvent) {
...
 
-  waitForRecordToFinish(Record);
-  MGraphBuilder.decrementLeafCountersForRecord(Record);
-  MGraphBuilder.cleanupCommandsForRecord(Record);
-  MGraphBuilder.removeRecordForMemObj(MemObj);
+    Lock.unlock();
+  }
+
+  {
+    // This only need a shared mutex as it only involves enqueueing and awaiting
```
Comment (sergey-semenov) [inline sycl/source/detail/scheduler/scheduler.cpp:174]:
```suggestion
    // This only needs a shared mutex as it only involves enqueueing and awaiting
```
Minor: typo

**Pull request merged**